clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -0,0 +1,221 @@
+pub struct UndocumentedUnsafeBlocks {
+    pub local_level: u32,
+}
+
+enum ErrorKind {
+    NoComment,
+    Placeholder
+}
+
+impl LateLintPass<'_> for UndocumentedUnsafeBlocks {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &'_ Block<'_>) {
+        if_chain! {
+            if self.local_level == 0;
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCKS, block.hir_id);
+            if !in_external_macro(cx.tcx.sess, block.span);
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                find_candidate(cx, block.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+
+    fn check_local(&mut self, cx: &LateContext<'_>, local: &'_ Local<'_>) {
+        if_chain! {
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCKS, local.hir_id);
+            if !in_external_macro(cx.tcx.sess, local.span);
+            if let Some(init) = local.init;
+            if let Some((block, element_count)) = find_block_in_expr(&init.kind);
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                self.local_level = self.local_level.saturating_add(element_count);
+
+                find_candidate(cx, local.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+
+    fn check_block_post(&mut self, _: &LateContext<'_>, _: &'_ Block<'_>) {
+        self.local_level = self.local_level.saturating_sub(1);
+    }
+}
+
+fn find_block_in_expr(expr_kind: &ExprKind<'hir>) -> Option<(&'tcx Block<'hir>, u32)> {
+    match expr_kind {
+        ExprKind::Array(elements) | ExprKind::Tup(elements) =>  {
+            let unsafe_blocks = elements.iter().filter_map(|e|
+                match e {
+                    Expr{
+                        kind: ExprKind::Block(block @ Block {
+                                                  rules: BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided),
+                                                  ..
+                                              }, _),
+                        ..
+                    } => Some(block),
+                    _ => None
+                }).collect::<Vec<_>>();
+
+            if let Some(block) = unsafe_blocks.get(0) {
+                return Some((block, unsafe_blocks.len().try_into().unwrap()))
+            }
+
+            None
+            },
+        ExprKind::Box(Expr {
+                          kind: ExprKind::Block(block, _),
+                          ..
+                      }) |
+        ExprKind::Unary(_, Expr {
+            kind: ExprKind::Block(block, _),
+            ..
+        }) |  ExprKind::Match(Expr {
+                                                            kind: ExprKind::Block(block, _),
```

Comment:
Is GitHub just displaying this weirdly or is the formatting totally broken?

Reply:
Yeah, I've been getting some weird results with `cargo dev fmt`.

Reply:
I wonder if this is a bug in rustfmt :thinking: Let's deal with this right before merging this PR.

Reply:
I just formatted it in playground and `cargo dev fmt --check` still passes. Is it just not doing anything then?

Reply:
I'm not sure. I'll test this myself before merging. Not sure what's going on.

Reply:
The check just failed on some missing semicolons, so its doing *something* at least.